### PR TITLE
Extract airgen

### PR DESCRIPTION
**Pull request by Schaeff**
I extracted the airgen extraction from #392

### CODE REVIEW

File: airgen/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    object::{Function, Instr, Link, LinkFrom, LinkTo, Location, Object, PILGraph},
+    parsed::{asm::InstructionBody, PilStatement},
+};
+
+const MAIN: &str = "Main";
+
+use number::FieldElement;
+
+pub fn compile<T: FieldElement>(input: AnalysisASMFile<T>) -> PILGraph<T> {
+    let main_location = Location::default().join("main");
```
Comment (leonardoalt) [inline airgen/src/lib.rs:18]: I think this should also be a constant

File: airgen/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        }
+    };
+
+    // get a list of all machines to instantiate. The order does not matter.
+    let mut queue = vec![(main_location.clone(), main_ty.clone())];
+
+    let mut instances = vec![];
+
+    while let Some((location, ty)) = queue.pop() {
+        let machine = input.machines.get(&ty).unwrap();
```
Comment (leonardoalt) [inline airgen/src/lib.rs:37]:
```suggestion
        let machine = input.machines[&ty];
```
If you're always unwrapping this should also work (just preference nit, fine with as is too)

File: airgen/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    // visit the tree compiling the machines
+    let objects = instances
+        .into_iter()
+        .map(|(location, ty)| {
+            let object = ASMPILConverter::convert_machine(&location, &ty, &input);
+            (location, object)
+        })
+        .collect();
+
+    let main_ty = input.machines.get(&main_ty).unwrap();
```
Comment (leonardoalt) [inline airgen/src/lib.rs:60]:
```suggestion
    let main_ty = input.machines.[&main_ty];
```
If you're always unwrapping this should also work (just preference nit, fine with as is too)

File: airgen/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        main: ast::object::Machine {
+            location: main_location,
+            function_id: main_ty.function_id.clone(),
+            latch: main_ty.latch.clone(),
+        },
+        entry_points: main_ty
+            .functions
+            .iter()
+            .map(|f| Function {
+                name: "main".to_string(),
```
Comment (leonardoalt) [inline airgen/src/lib.rs:72]: should be a constant (used twice now)

File: airgen/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    fn convert_machine(
+        location: &'a Location,
+        ty: &'a str,
+        input: &'a AnalysisASMFile<T>,
+    ) -> Object<T> {
+        Self::new(location, input).convert_machine_inner(ty)
+    }
+
+    fn convert_machine_inner(mut self, ty: &str) -> Object<T> {
+        let input = self.machines.get(ty).unwrap().clone();
```
Comment (leonardoalt) [inline airgen/src/lib.rs:113]:
```suggestion
        let input = self.machines[ty].clone();
```

File: airgen/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    fn handle_instruction_def(
+        &mut self,
+        InstructionDefinitionStatement {
+            start: _,
+            name,
+            params,
+            body,
+        }: InstructionDefinitionStatement<T>,
+    ) -> Option<Link<T>> {
+        let instruction_flag = format!("instr_{name}");
```
Comment (leonardoalt) [inline airgen/src/lib.rs:145]: Just to confirm, this doesn't require that the input has this column, but rather generates it here, right?

File: airgen/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+                // get the machine type name for this submachine from the submachine delcarations
+                let instance_ty_name = self
+                    .submachines
+                    .iter()
+                    .find(|s| s.name == instance)
+                    .unwrap()
+                    .ty
+                    .clone();
+                // get the machine type from the machine map
+                let instance_ty = self.machines.get(&instance_ty_name).unwrap();
```
Comment (leonardoalt) [inline airgen/src/lib.rs:164]:
```suggestion
                let instance_ty = self.machines[&instance_ty_name];
```

File: linker/src/lib.rs [hunk 8]
```diff
@@ -132,14 +142,20 @@ mod test {
...
+        ))
     }
 
     #[test]
     fn degree() {
         // a graph with two objects of degree `main_degree` and `foo_degree`
         let test_graph = |main_degree, foo_degree| PILGraph {
-            main: Location::from("main".to_string()),
+            main: Machine {
+                location: Location::from("main".to_string()),
```
Comment (leonardoalt) [inline linker/src/lib.rs:155]: Should be a constant

File: airgen/src/lib.rs (hunk 4 above)
Comment (Schaeff) [reply airgen/src/lib.rs:72]: will fix

File: airgen/src/lib.rs (hunk 6 above)
Comment (Schaeff) [reply airgen/src/lib.rs:145]: will add todo, it will be fixed once we have CM links

**Pull request merged**